articles/quickstart/native/ios-swift-siwa/_includes/_login_centralized.md

```diff
@@ -0,0 +1,306 @@
+* Use `SimpleKeychain` to store the Apple user ID
+* Check the credential state using Apple's APIs before retrieving Auth0 credentials
+
+### Store credentials
+
+The credentials manager retrieves stored credentials from the keychain and checks if the Access Token is still valid:
+
+* If the current credentials are still valid, the credentials manager returns them
+* If the Access Token has expired, the credentials manager renews them using the Refresh Token and returns them
+
+At the top of `ViewController.swift`, add a `CredentialsManager` and `SimpleKeychain` as follows:
+
+```swift
+let credentialsManager = CredentialsManager(authentication: Auth0.authentication())
+let keychain = A0SimpleKeychain(service: "Auth0")
+
+// Add a place to store credentials locally once they're renewed
+var credentials: Credentials?
+```
+
+Next, modify the token exchange call you added earlier to store credentials when the user logs in:
+
+```swift
+Auth0
+    .authentication()
+    .tokenExchange(withAppleAuthorizationCode: authCode).start { result in
+        switch(result) {
+        case .success(let credentials):
+
+            // NEW - store the credentials in the credentials manager
+            self.credentialsManager.store(credentials: credentials)
+
+        case .failure(let error):
+            print("Exchange Failed: \(error)")
+        }
+}
+```
+
+### Renew authentication state
+
+Add a function that tries to renew the user's login session.
+
+```swift
+func tryRenewAuth(_ callback: @escaping (Credentials?, Error?) -> ()) {
+    let provider = ASAuthorizationAppleIDProvider()
+        
+    // Try to fetch the user ID
+    guard let userID = keychain.string(forKey: "userId") else {
+        return callback(nil, nil)
+    }
+
+    // Check the Apple credential state
+    provider.getCredentialState(forUserID: userID) { state, error in
+        switch state {
+        case .authorized:
+            // Try to get credentials from the creds manager (ID token, Access Token, etc)
+            self.credentialsManager.credentials { error, credentials in
+                guard error == nil, let credentials = credentials else {
+                    return callback(nil, error)
+                }
+                
+                callback(credentials, error)
+            }
+            
+        default:
+            // User is not authorized - clear credentials
+            
+            self.keychain.deleteEntry(forKey: "userId")
+            self.credentialsManager.clear()
+            
+            callback(nil, error)
+        }
+    }
+}
+```
+
+:::note
+Calling `credentialsManager.credentials` _automatically renews_ the Access Token if it has expired, using the refresh token. It is important to ensure this call only executes if `getCredentialState` returns `authorized`, so that you can be sure the refresh token is being used by an authorized user. Otherwise, the credentials must be cleared and the login session thrown away.
```

Comment:
Calling `credentialsManager.credentials` _automatically renews_ the Access Token if it has expired, using the refresh token. This call should only execute if `getCredentialState` returns `authorized`, so the refresh token is only used by an authorized user. Otherwise, the credentials must be cleared and the login session thrown away.